Improvements to Compiler Compatibility and CI/CD Improvements

ashao opened the pull request:
This PR includes a number of changes that expands the testing coverage and enables SmartRedis to be built with the Nvidia HPC compiler suite.

- The Nvidia HPC toolchain was not quite compiling correctly because some of our dependencies implicitly assume that GCC is being used to compile. This was causing spurious errors do to the fact that these warnings that were interpreted as errors within the dependencies' build steps. This was solved by enforcing that all dependencies are compiled using the system's GCC (and can be overwritten by the user by setting DEPS_CC and DEPS_CXX). Additionally this adds the Nvidia compiler chain to the Github Actions CI/CD.
- Bumping the GCC versions also revealed some innocuous but standard-violating parts of the SmartRedis codebase (e.g. assigning negative values to unsigned integers, unreachable codeblocks)
- The version of Redis is now consistent with the version we use with SmartSim
- The container version was bumped to use Ubuntu 22.04 in order to start testing GCC 12
- The installation of the Intel compilers (and other third party dependencies) via apt was updated to conform with each provider's current instructions
- The version of Torch for testing is now pinned to 1.11.0, the version that SmartSim currently supports

billschereriii commented on .github/workflows/run_tests.yml:
```diff
@@ -84,39 +86,58 @@ jobs:
+            | sudo tee /etc/apt/sources.list.d/oneAPI.list &&
+          sudo apt update -y && \
+          sudo apt install -y intel-oneapi-compiler-fortran intel-oneapi-compiler-dpcpp-cpp \
+            intel-oneapi-mpi intel-oneapi-mpi-devel intel-oneapi-mkl &&
```
This is the preferred way to install the intel toolchain in CI/CD now?

ashao replied:
Yes it also results in a pretty minimal installation

billschereriii commented on .github/workflows/run_tests.yml:
```diff
@@ -58,13 +58,15 @@ jobs:
-        py_v: ['3.7.x', '3.8.x', '3.9.x', '3.10.x'] # versions of Python
-        compiler: [intel, 8, 9, 10, 11] # intel compiler, and versions of GNU compiler
+        py_v: ['3.8.x', '3.9.x', '3.10.x'] # versions of Python
+        compiler: [nvhpc-23-11, intel, gcc-11, gcc-12] # intel compiler, and versions of GNU compiler
```
We're identifying other vendors' toolchains by vendor name, not product name. Might make subsequent code more readable if it were to say "nvidia"

billschereriii commented on .github/workflows/run_tests.yml:
```diff
@@ -58,13 +58,15 @@ jobs:
     env:
-      FC: gfortran-${{ matrix.compiler }}
-      GCC_V: ${{ matrix.compiler }} # used when the compiler is gcc/gfortran
+      COMPILER: ${{ matrix.compiler }} # used when the compiler is gcc/gfortran
```
Are we going to need to differentiate like this for NVIDIA as well?

ashao replied:
The specification of these versions is now done within the setup section of each compiler chain. The only one that doesn't this to be done is the NVHPC compiler chain because it is installed as a single package (and not individually like Intel and GCC).

billschereriii commented on .github/workflows/run_tests.yml:
```diff
@@ -146,52 +168,23 @@ jobs:
-          sudo mkdir -p /usr/lib/redis/modules/ &&
-          sudo docker cp $CONTAINER_NAME:/usr/lib/redis/modules/redisai.so /usr/lib/redis/modules &&
-          sudo docker cp $CONTAINER_NAME:/usr/lib/redis/modules/backends/ /usr/lib/redis/modules/ &&
+          make test-deps &&
```
I didn't do this before because it is slower than grabbing from the docker container. As it is, explicitly calling make on test-deps is redundant since it will occur as needed with any `make test*` step

ashao replied:
The container that we were extracting RedisAI from uses Redis 6, whereas SmartSim is currently using Redis 7. Thus, these modules are now inconsistent with the redis-server that we build. While we could potentially continue testing with Redis 6, upgrading to Redis 7 ensures consistency with SmartSim

billschereriii commented on .github/workflows/run_tests.yml:
```diff
@@ -146,52 +168,23 @@ jobs:
-      - name: Test static build -- intel compilers
-        if: contains(matrix.os, 'ubuntu') && contains(matrix.compiler, 'intel')
-        run: make build-example-serial SR_LINK=Static SR_FORTRAN=ON
+          echo "$PWD/third-party/redis/src" >> $GITHUB_PATH
```
Why is this needed? The test drivers know how to find redis utils from the SR root

ashao replied:
Good point I'll track this down

ashao replied:
This was indeed not needed

billschereriii commented on .github/workflows/run_tests.yml:
```diff
@@ -146,52 +168,23 @@ jobs:
+      - name: Run examples
+        run: |
+          make test-examples SR_FORTRAN=ON SR_PYTHON=ON SR_LINK=$LINK_TYPE \
+          SR_TEST_PORT=7000 SR_TEST_REDISAI_VER=v${{ matrix.rai_v }}
```
I didn't do this before so that we could have the examples and static builds run in parallel with the main tests. Adding them here extends the time for each runner. (I didn't look to see how long the runs are taking now though, so maybe it's not so bad.)

I know we were only running the examples with some compiler chains previously. This may be the source of some of your failures

billschereriii commented on .github/workflows/run_tests.yml:
```diff
@@ -58,13 +58,15 @@ jobs:
-        compiler: [intel, 8, 9, 10, 11] # intel compiler, and versions of GNU compiler
+        py_v: ['3.8.x', '3.9.x', '3.10.x'] # versions of Python
+        compiler: [nvhpc-23-11, intel, gcc-11, gcc-12] # intel compiler, and versions of GNU compiler
+        link_type: [Static, Shared]
```
This doubles the number of runners needed, which in turn will make the CI/CD process take a much longer amount of time to complete since it will be pushing us beyond the available amount

ashao replied:
Good point, will follow up with the rest of the group

billschereriii commented on Makefile:
```diff
@@ -85,7 +87,7 @@ help:
 # help: SR_BUILD {Release, Debug, Coverage} -- optimization level for the build
 # help: SR_LINK {Shared, Static} -- linkage for the SmartRedis library
-# help: SR_PEDANTIC {OFF, ON} -- GNU only; enable pickiest compiler settings
+# help: SR_PEDANTIC {OFF, ON} -- GNU only; enable pickiest compiler settings, currently fails do to warnings
```
do => due

billschereriii commented on Makefile:
```diff
@@ -216,7 +218,7 @@ build-example-serial: lib
 	@cmake -S examples/parallel -B build/$(SR_BUILD)/examples/$(SR_LINK)/parallel \
 		-DSR_BUILD=$(SR_BUILD) -DSR_LINK=$(SR_LINK) -DSR_FORTRAN=$(SR_FORTRAN)
-	@cmake --build build/$(SR_BUILD)/examples/$(SR_LINK)/parellel
+	@cmake --build build/$(SR_BUILD)/examples/$(SR_LINK)/parallel
```
Yikes!

billschereriii commented on Makefile:
```diff
@@ -44,6 +44,8 @@ CATCH2_URL := https://github.com/catchorg/Catch2.git
 LCOV_URL := https://github.com/linux-test-project/lcov.git
 LCOV_VER := v2.0
+DEP_CC := gcc
+DEP_CXX := g++
```
If these are used only for third party dependencies, add a comment here to explain that

billschereriii commented on setup.cfg:
```diff
@@ -44,7 +44,7 @@ dev =
     isort==5.6.4
     pylint>=2.10.0
-    torch>=1.7.1
+    torch==1.11.0
```
This feels unrelated to the rest of the changes here

billschereriii commented on src/cpp/client.cpp:
```diff
@@ -1226,7 +1226,7 @@ void Client::set_data_source(std::string source_id)
     bool valid_prefix = false;
     size_t num_prefix = _get_key_prefixes.size();
-    size_t save_index = -1;
+    size_t save_index = 0;
```
If this is a bug it should be in a separate PR

billschereriii commented on src/cpp/dbnode.cpp:
```diff
@@ -32,7 +32,7 @@ using namespace SmartRedis;
 // DBNode constructor
 DBNode::DBNode()
-    : address(), name(""), lower_hash_slot(-1), upper_hash_slot(-1)
+    : address(), name(""), lower_hash_slot(0), upper_hash_slot(0)
```
This should be a separate PR (along with the previous?)

billschereriii commented on src/cpp/metadata.cpp:
```diff
@@ -230,11 +230,9 @@ void MetaData::get_scalar_values(const std::string& name,
         case SRMetadataTypeString:
             throw SRRuntimeException("MetaData.get_scalar_values() "\
                                      "requested invalid MetaDataType.");
-            break;
```
I'm assuming that these changes are to remove compiler warnings for unreachable code. Should they all be in a separate PR?

billschereriii commented on utils/launch_redis.py:
```diff
@@ -37,15 +37,10 @@ def check_availability(n_nodes, port, udsport):
+    rediscli = (
+        pathlib.Path(__file__).parent.parent
+        / "third-party/redis/src/redis-cli"
+    ).resolve()
```
So again, this is removing a performance optimization that we intentionally left in. While I don't object, it does seem like it should be a separate PR

billschereriii commented on utils/launch_redis.py:
```diff
@@ -70,17 +65,11 @@ def stop_db(n_nodes, port, udsport):
+    rediscli = (
+        pathlib.Path(__file__).parent.parent
+        / "third-party/redis/src/redis-cli"
+    ).resolve()
```
Same as previous

billschereriii commented on utils/launch_redis.py:
```diff
@@ -159,43 +148,30 @@ def create_db(n_nodes, port, device, rai_ver, udsport):
+        pathlib.Path(__file__).parent.parent
+        / "third-party/redis/src/redis-server"
+    ).resolve()
+    rediscli = os.path.dirname(redisserver) + "/redis-cli"
```
and again here

billschereriii reviewed (changes requested):
This is at least four PRs rolled together into one, which makes it a bit tough to review:

1. NVIDIA compiler support
2. Compiler warning optimization
3. Intel compiler installation in CI/CD
4. CI/CD changes to eliminate RedisAI installation from Docker container and to move static build and examples into the main "always test"

Changes 1-3 individually look fine, I just think they should have been in separate PRs.
Change 4, however, is probably extending the CI/CD time on commit by over a factor of 2 (though to be fair I haven't actually looked to confirm that this is the case). There probably needs to be a group discussion as to whether that's acceptable in view of expected incoming development.

Additional comments:

1. Not all tests are currently passing. This needs to be fixed before proceeding
2. The change log is not yet updated

billschereriii commented on Makefile:
```diff
@@ -518,10 +522,10 @@ install/lib/libredis++.a: third-party/redis-plus-plus
+	cmake -DCMAKE_BUILD_TYPE=Release -DREDIS_PLUS_PLUS_BUILD_TEST=OFF \
+		  -DREDIS_PLUS_PLUS_BUILD_SHARED=OFF -DCMAKE_PREFIX_PATH="../../../install/lib/" \
+		  -DCMAKE_INSTALL_PREFIX="../../../install" -DCMAKE_CXX_STANDARD=17 \
+		  -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_C_COMPILER=$(DEP_CC) -DCMAKE_CXX_COMPILER=$(DEP_CXX) .. && \
```
I can't remember why this was done in a subshell previously

ashao commented:
Thanks @billschereriii for your careful feedback. I have addressed your comments (some of which we discussed offline). One outstanding issue is the reduced breadth of the Github Actions. I did try to expand the matrix as a post-merge action, however previous versions of the Intel compiler set up the compiler environment differently. This requires a non-trivial solution, so will be addressing this in a future PR.

billschereriii commented on doc/changelog.rst:
```diff
@@ -8,16 +8,30 @@ To be released at some future point in time
 
 Description
 
+- Tweak the build system to enable building SmartRedis with Nvidia's toolchain
```
Nvidia's HPC toolchain?

billschereriii commented on doc/changelog.rst:
```diff
@@ -8,16 +8,30 @@ To be released at some future point in time
 
+- Tweak the build system to enable building SmartRedis with Nvidia's toolchain
+- Improvements/upgrades to the container used for Github actions
+- Rework code to avoid compiler warnings
```
Code updates to resolve warnings?

billschereriii commented on doc/changelog.rst:
```diff
@@ -8,16 +8,30 @@ To be released at some future point in time
 
+- Change the order of arguments in our MakeFile to ensure that all dependencies are compiled with GCC (PR448_)
+- Add new user-configurable parameters DEP_CC, DEP_CXX to control which compiler is used to build dependencies (PR448_)
+- Bump the container version to Ubuntu 22.04 to be able to start testing GCC 12 (PR448_)
```
This should say CI/CD in it somewhere

ashao replied:
Agreed: Prefixed these with CI/CD (since most people's eyes will just glaze over)

billschereriii commented on doc/changelog.rst:
```diff
@@ -8,16 +8,30 @@ To be released at some future point in time
+- Change the order of arguments in our MakeFile to ensure that all dependencies are compiled with GCC (PR448_)
+- Add new user-configurable parameters DEP_CC, DEP_CXX to control which compiler is used to build dependencies (PR448_)
+- Bump the container version to Ubuntu 22.04 to be able to start testing GCC 12 (PR448_)
+- Bump the versions of GCC used to only the currently maintained versions (PR448_)
```
Restrict? And this should say CI/CD in it to keep a customer from thinking we're not letting them use gcc8 any more

billschereriii commented on doc/changelog.rst:
```diff
@@ -8,16 +8,30 @@ To be released at some future point in time
+- Compile Redis and RedisAI and use those versions in testing instead of extracting from a container (PR448_)
+- Bump the version of Redis used in testing to 7.0.5, the same version as we use with SmartSim (PR448_)
+- Pin the Torch version to 1.11.0, the same as supported in SmartSim (PR448_)
+- Ameliorate some compiler warnings related that were flagged in GCC 12 (unreachable code blocks, invalid values for the type) (PR448_)
```
suggest replacing "invlaid values for the type" with "signed/unsigned mismatches"

billschereriii commented on doc/install/lib.rst:
```diff
@@ -25,6 +25,24 @@ installed in ``SmartRedis/install/include/``.  The library installation can be
+    # GCC Toolchain
+    make lib-with-fortran CC=gcc CXX=g++ FC=gfortran
+
+    # Nvidia Toolchain
```
Nvidia (HPC) Toolchain

ashao replied:
Agreed, I did put in a table instead

billschereriii commented on doc/installation.rst:
```diff
@@ -33,6 +33,7 @@ SmartRedis is tested with the following compilers on a daily basis:
     * - GNU (GCC/GFortran)
     * - Intel oneAPI (icx/icpx/ifort)
+    * - NVHPC
     * - Apple Clang
```
Might need to move Clang to the less frequent section now

ashao replied:
I don't want to remove this yet, since it still does compile. Plus, I think because the container isn't used anymore, we might be able to restore the Apple Clang compiler chain.

billschereriii reviewed (approved):
Several opportunities for textual clarification, but I have full confidence in your ability to review these appropriately.

Thanks for taking on these changes. It will be a definite improvement to the library to have them in!

I do still recommend conferring with the group on the topic of which compilers to include or not include in CI/CD testing.
